File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/MapOfStringsToStringConverter.java
package openfoodfacts.github.scrachx.openfood.models.entities;

import static android.util.Base64.DEFAULT;

import android.util.Base64;
import android.util.Log;

import androidx.annotation.NonNull;

import org.greenrobot.greendao.converter.PropertyConverter;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.HashMap;
import java.util.Map;

class MapOfStringsToStringConverter implements PropertyConverter<Map<String, String>, String> {
    private static final String LOG_TAG;

    @Override
    @NonNull
    public Map<String, String> convertToEntityProperty(String databaseValue) {
        if (databaseValue == null) {
            return new HashMap<>();
        }

        final Map<String, String> decodedResult = new HashMap<>();
        try {
            ByteArrayInputStream bis = new ByteArrayInputStream(Base64.decode(databaseValue, DEFAULT));
            ObjectInputStream objectInputStream = new ObjectInputStream(bis);
            @SuppressWarnings("UNCHECKED_CAST")
            final Map<String, String> readMap = (Map<String, String>) objectInputStream.readObject() ;
            decodedResult.putAll(readMap);
        } catch (IOException | ClassNotFoundException e) {
            Log.e(LOG_TAG, "Cannot serialize map to database.", e);
        }
        return decodedResult;
    }

    @Override
    @NonNull
    public String convertToDatabaseValue(Map<String, String> entity) {
        HashMap<String, String> map = entity != null ? new HashMap<>(entity) : new HashMap<>();

        try {
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            ObjectOutputStream objectOutputStream = new ObjectOutputStream(bos);
            objectOutputStream.writeObject(map);
            objectOutputStream.flush();
            return Base64.encodeToString(bos.toByteArray(), DEFAULT);
        } catch (IOException e) {
            Log.e(LOG_TAG, "Cannot serialize map to database.", e);
        }
        return "";
    }

    static {
        LOG_TAG = MapOfStringsToStringConverter.class.getSimpleName();
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/MapOfStringsToStringConverter.kt
package openfoodfacts.github.scrachx.openfood.models.entities

import android.util.Base64
import android.util.Log
import org.greenrobot.greendao.converter.PropertyConverter
import java.io.*
import java.util.*

class MapOfStringsToStringConverter : PropertyConverter<Map<String, String>?, String?> {
    override fun convertToEntityProperty(databaseValue: String?): Map<String, String>? {
        if (databaseValue == null) return null
        try {
            return ByteArrayInputStream(Base64.decode(databaseValue, Base64.DEFAULT)).use { bis ->
                ObjectInputStream(bis).use {
                    it.readObject() as HashMap<String, String>
                }
            }
        } catch (e: Exception) {
            Log.e(LOG_TAG, "Cannot deserialize map from database.", e)
        }
        return null
    }

    override fun convertToDatabaseValue(entityProperty: Map<String, String>?): String? {
        if (entityProperty == null) return null
        try {
            return ByteArrayOutputStream().use { bos ->
                ObjectOutputStream(bos).use {
                    it.writeObject(HashMap(entityProperty))
                    it.flush()
                    Base64.encodeToString(bos.toByteArray(), Base64.DEFAULT)
                }
            }
        } catch (e: IOException) {
            Log.e(LOG_TAG, "Cannot serialize map to database.", e)
        }
        return null
    }

    companion object {
        val LOG_TAG = this::class.simpleName
    }
}